File: src/cli/commands/inspect.c
#include "commands.h"
#include "nurl_utils.h"
#include "errors/nurl_diag.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>

int nurl_cmd_inspect(const char *url, const CommonArgs *common) {
    char *scheme = NULL;
    char *host = NULL;
    char *path = NULL;
    int port = 0;

    if (nurl_utils_parse_url(url, &scheme, &host, &port, &path) != 0) {
        nurl_diag_err("malformed URL '%s' provided for inspection.", url);
        nurl_diag_hint("ensure the URL uses a supported scheme like 'https://' and has a valid hostname.");
        return NURL_ERR_INVALID_URL;
    }

    // Print Request Line and Host
    printf("> %s %s HTTP/1.1\n", common->method ? common->method : "GET", path);
    printf("> Host: %s\n", host);
    printf("> User-Agent: nurl/" NURL_VERSION "\n");
    printf("> Connection: close\n");

    // 1. User specified headers
    for (size_t i = 0; i < common->header_count; i++) {
        char *line = strdup(common->header[i]);
        if (line) {
            char *colon = strchr(line, ':');
            if (colon) {
                *colon = '\0';
                char *key = line;
                char *val = colon + 1;
                while (*val && isspace((unsigned char)*val)) val++;
                const char *redacted = nurl_utils_redact_header(key, val);
                printf("> %s: %s\n", key, redacted);
            } else {
                printf("> %s\n", common->header[i]);
            }
            free(line);
        }
    }

    // 2. Auth Header
    if (!common->no_auth) {
        if (common->bearer || common->token) {
            if (!nurl_utils_has_header(common->header, common->header_count, "Authorization")) {
                printf("> Authorization: [hidden]\n");
            }
        } else if (common->user) {
            if (!nurl_utils_has_header(common->header, common->header_count, "Authorization")) {
                printf("> Authorization: [hidden]\n");
            }
        }
    }

    // 3. JSON Header
    if (common->json && !nurl_utils_has_header(common->header, common->header_count, "Content-Type")) {
        printf("> Content-Type: application/json\n");
    }

    // 4. Content-Length Header
    size_t body_len = 0;
    if (common->data) {
        body_len = common->data_len > 0 ? common->data_len : strlen(common->data);
    }
    if (body_len > 0 && !nurl_utils_has_header(common->header, common->header_count, "Content-Length")) {
        printf("> Content-Length: %zu\n", body_len);
    }

    printf(">\n");

    // Print Body
    if (common->data && body_len > 0) {
        fwrite(common->data, 1, body_len, stdout);
        printf("\n");
    }

    free(scheme);
    free(host);
    free(path);
    return 0;
}

File: src/cli/commands/inspect.rs
use clap::Args;
use crate::cli::parser::CommonArgs;
use crate::cli::error_handler::print_and_exit;
use crate::utils::body::BodySource;

#[derive(Args, Debug)]
pub struct InspectArgs {
    /// HTTP Method (GET, POST, PUT, etc.)
    pub method: String,

    /// Target URL
    pub url: String,

    /// Request body data
    #[arg(short = 'd', long)]
    pub body: Option<String>,

    #[clap(flatten)]
    pub common: CommonArgs,
}

pub fn run(args: InspectArgs) {
    let clean_url = crate::utils::url::clean_url(&args.url);
    let mut headers = match crate::utils::headers::parse_headers(&args.common.header) {
        Ok(h) => h,
        Err(e) => {
            if !args.common.silent {
                print_and_exit(&e);
            } else {
                std::process::exit(e.exit_code());
            }
        }
    };

    let has_content_type = headers.iter().any(|(k, _)| k.to_lowercase() == "content-type");
    if args.common.json && !has_content_type {
        headers.push(("Content-Type".to_string(), "application/json".to_string()));
    }

    if !args.common.no_auth {
        if let Some(ref tok) = args.common.bearer.as_ref().or(args.common.token.as_ref()) {
            headers.push(("Authorization".to_string(), format!("Bearer {}", tok)));
        } else if let Some(ref user) = args.common.user {
            use base64::Engine;
            let encoded = base64::engine::general_purpose::STANDARD.encode(user.as_bytes());
            headers.push(("Authorization".to_string(), format!("Basic {}", encoded)));
        }
    } else {
        headers.retain(|(k, _)| k.to_lowercase() != "authorization");
    }

    let body_source = match args.body {
        Some(s) => BodySource::Raw(s),
        None => BodySource::None,
    };

    let body = match body_source.resolve() {
        Ok(b) => b,
        Err(e) => {
            if !args.common.silent {
                print_and_exit(&e);
            } else {
                std::process::exit(e.exit_code());
            }
        }
    };

    let has_content_length = headers.iter().any(|(k, _)| k.to_lowercase() == "content-length");
    if let Some(ref b) = body {
        if !has_content_length {
            headers.push(("Content-Length".to_string(), b.len().to_string()));
        }
    }

    let parsed_url = url::Url::parse(&clean_url).ok();
    let host = parsed_url.as_ref().and_then(|u| u.host_str()).unwrap_or(&clean_url);
    let path_and_query = parsed_url.as_ref()
        .map(|u| {
            let mut pq = u.path().to_string();
            if let Some(q) = u.query() {
                pq.push('?');
                pq.push_str(q);
            }
            pq
        })
        .unwrap_or_else(|| "/".to_string());

    println!("> {} {} HTTP/1.1", args.method.to_uppercase(), path_and_query);
    println!("> Host: {}", host);
    for (k, v) in &headers {
        let redacted = crate::utils::headers::redact_header_value(k, v);
        println!("> {}: {}", k, redacted);
    }
    println!(">");
    if let Some(ref b) = body {
        let body_str = String::from_utf8_lossy(b);
        for line in body_str.lines() {
            println!("> {}", line);
        }
    }
}
